src/typecheck.rs

```diff
@@ -502,16 +539,16 @@ fn type_check_(
 
     match t.as_ref() {
         // null is inferred to be of type Dyn
-        Term::Null => unify(state, strict, ty, mk_typewrapper::dynamic())
+        Term::Null => check_sub(state, strict, mk_typewrapper::dynamic(), ty)
```

Comment:
I hesitated on this one. The duplication of typing rules (infer mode and checking mode) leads to actual code duplication: in most cases, for example in the case of checking a literal against a ground type, checking is literally the same thing as inferring the type and then constraining it to be a subtype of the checked type. So mixing both is a way of sharing more common code.

However, it does make the code more entangled. And in one case, inferring and constraining is less general than checking (for heterogeneous lists, where there's a special casing, which in fact corresponds to implementing proper checking instead of piggybacking on this infer + constrain).